File: Java/325. Maximum Size Subarray Sum Equals k.java
class Solution {
    public int maxSubArrayLen(int[] nums, int k) {
        Map<Integer, Integer> cumu_sum = new HashMap<>();
        cumu_sum.put(0, -1);
        int res = 0, sub_sum = 0;
        
        for(int i=0; i<nums.length; i++){
            sub_sum += nums[i];
            int comp = sub_sum - k;
            
            if(cumu_sum.containsKey(comp)){
                res = Math.max(res, i-cumu_sum.get(comp));
            }
            
            if(!cumu_sum.containsKey(sub_sum)){
                cumu_sum.put(sub_sum, i);
            }
        }
        
        return res;
    }
}

File: Python3/325. Maximum Size Subarray Sum Equals k.py
class Solution:
    def maxSubArrayLen(self, nums: List[int], k: int) -> int:
        # sum_subarray[i:j] = sum([:j]) - sum([:i])
        # dict: key -> cumulative sum from start to this index, value -> indices
        cumu_sum = {}
        
        # add pseudo sum 0, such that sub_sum == k can be counted
        cumu_sum[0] = -1
        
        sub_sum = 0
        res = 0
        
        # cur = sum, find (sum-k)
        # if it's in cumu_sum, update res = max(res, cur_i-pre_i)
        for i, n in enumerate(nums):
            sub_sum += n
            if sub_sum not in cumu_sum:
                cumu_sum[sub_sum] = i
            
            if (sub_sum - k) in cumu_sum:
                res = max(res, i - cumu_sum[sub_sum-k])
                
        return res